ui/source/js/components/game.js

```diff
@@ -1,78 +1,127 @@
 import React, {Component} from 'react';
 import Board from './board.js';
+import Stats from './stats.js';
+import Controls from './controls.js';
+import getWordlist from '../helpers/generateWordList.js';
+import getOwnershipList from '../helpers/generateOwners.js';
+import generateTeams from '../helpers/teams.js';
+import switchTurns from '../helpers/switchTurns.js';
+import keepScore from '../helpers/keepScore.js';
 import calculateWinner from '../helpers/winner.js';
 
+const gamePiecesCount = 25;
+const blackPieceCount = 1;
+const bluePiecesCount = 9;
+const redPiecesCount = 8;
+
+const strResetGame = 'Start a new game?';
+
+function buildPieceArray() {
+	const wordArray = getWordlist(gamePiecesCount);
+	const ownershipArray = getOwnershipList(gamePiecesCount, blackPieceCount, bluePiecesCount, redPiecesCount);
+	const piecesArray = [];
+	for (let i = 0; i < gamePiecesCount; i++) {
+		const piece = {
+			word: wordArray[i],
+			owner: ownershipArray[i],
+			isChosen: false
+		}
+		piecesArray.push(piece);
+	}
+	return piecesArray;
+}
+
 class Game extends React.Component {
 	constructor(props) {
 		super(props);
 		this.state = {
-			history: [{
-				pieces: Array(9).fill(null)
-			}],
-			xIsNext: true,
-			stepNumber: 0
+			isBlueTurn: true,
+			score: {
+				black: blackPieceCount,
+				blue: bluePiecesCount,
+				red: redPiecesCount
+			},
+			pieces: buildPieceArray(),
+			viewer: 0,
+			winner: null,
+			teams: generateTeams()
 		};
 	}
 
-	handleClick(i) {
-		const history = this.state.history.slice(0, this.state.stepNumber + 1);
-		const current = history[history.length - 1];
-		const pieces = current.pieces.slice();
-		if (calculateWinner(pieces) || pieces[i]) {
-			return;
-		}
-		pieces[i] = this.state.xIsNext ? 'X' : 'O';
-		this.setState({
-			history: history.concat([{
-				pieces: pieces
-			}]),
-			stepNumber: history.length,
-			xIsNext: !this.state.xIsNext
-		});
+	switchTeamsClick() {
+		this.setState({ isBlueTurn: !this.state.isBlueTurn });
 	}
 
-	jumpTo(step) {
-		this.setState({
-			stepNumber: step,
-			xIsNext: (step % 2) === 0,
-		});
+	setViewer(i) {
+		this.setState({ viewer: i });
 	}
 
-	render() {
-		const history = this.state.history;
-		const current = history[this.state.stepNumber];
-		const winner = calculateWinner(current.pieces);
+	pieceChosenClick(i) {
+		if (this.state.winner === null) {
+			const updatedPieces = Array.from(this.state.pieces);
+			const clickedPieceOwner = updatedPieces[i].owner;
+			updatedPieces[i].isChosen = true;
+			const newScore = keepScore(updatedPieces, blackPieceCount, bluePiecesCount, redPiecesCount);
+			const newWinner = calculateWinner(newScore, this.state.isBlueTurn);
+			this.setState({
+				score: newScore,
+				pieces: updatedPieces,
+				winner: newWinner,
+				isBlueTurn: switchTurns(this.state.isBlueTurn, clickedPieceOwner)
+			});
+			if (newWinner !== null) {
+				this.endGame(newWinner);
```

Comment:
It's possible you could end up with a race condition, since `setState()` is asynchronous. Your `setState()` on line 67 might not complete before your `setState()` in your `endGame()` function fires. So, it's possible that your `endGame()` function could be updating state with outdated `pieces`. Whether or not this would cause an actual bug in your game would require further investigation, but it would be a good idea to combine two back-to-back `setState()` calls into one call if possible.

Reply:
Good tip on it being Async. Thanks! I'll refactor that section.